File: dotnet/Allmark/RenderConsole/ConsoleTableRenderer.cs
namespace Allmark.Render;

using Allmark.Types;
using System.Text;

public static class ConsoleTableRenderer
{
    public static OutputRenderer Create()
    {
        return new OutputRenderer
        {
            Name = "table",
            Render = (node, state, decode) => Render(node, state),
        };
    }

    public static void Render(MarkdownNode node, RendererState state)
    {
        var style = Ansi.Dim;

        if (node.Children == null || node.Children.Count == 0)
        {
            return;
        }

        var headerRow = node.Children[0];
        var dataRows = node.Children.Skip(1).ToList();

        var headerCells = headerRow.Children ?? new List<MarkdownNode>();
        var cellTexts = new List<List<string>>();

        var maxColumns = Math.Max(headerCells.Count, dataRows.Max(r => r.Children?.Count ?? 0));
        var columnWidths = Enumerable.Repeat(0, maxColumns).ToList();
        var alignments = Enumerable.Repeat("", maxColumns).ToList();

        for (int i = 0; i < headerCells.Count; i++)
        {
            var text = GetTextFromNode(headerCells[i]);
            if (cellTexts.Count == 0)
            {
                cellTexts.Add(new List<string>());
            }
            cellTexts[0].Add(text);
            columnWidths[i] = Math.Max(columnWidths[i], text.Length + 2);
            alignments[i] = headerCells[i].Info ?? "";
        }

        for (int r = 0; r < dataRows.Count; r++)
        {
            var row = dataRows[r];
            var rowCells = row.Children ?? new List<MarkdownNode>();
            if (cellTexts.Count <= r + 1)
            {
                cellTexts.Add(new List<string>());
            }
            for (int c = 0; c < rowCells.Count; c++)
            {
                var text = GetTextFromNode(rowCells[c]);
                while (cellTexts[r + 1].Count <= c)
                {
                    cellTexts[r + 1].Add("");
                }
                cellTexts[r + 1][c] = text;
                if (c < columnWidths.Count)
                {
                    columnWidths[c] = Math.Max(columnWidths[c], text.Length + 2);
                }
            }
        }

        string MakeLine(string left, string mid, string right, string sep, int[] widths)
        {
            var line = new StringBuilder();
            line.Append(left);
            for (int i = 0; i < widths.Length; i++)
            {
                line.Append(new string('─', widths[i]));
                if (i < widths.Length - 1)
                {
                    line.Append(i == 0 ? mid : sep);
                }
            }
            line.Append(right);
            return $"{style}{line}{Ansi.Reset}\n";
        }

        string[] alignmentsArray = alignments.ToArray();

        string[][] cellTextsArray = cellTexts.Select(r => r.ToArray()).ToArray();
        int[] targetWidths = columnWidths.ToArray();
        string[][][]? wrappedCells = null;

        int totalWidth = targetWidths.Sum() + maxColumns + 1;
        if (state.LineWidth.HasValue && totalWidth > state.LineWidth.Value)
        {
            targetWidths = FitColumns(targetWidths, state.LineWidth.Value, maxColumns, cellTextsArray);
            wrappedCells = WrapAllCells(cellTextsArray, targetWidths);
            targetWidths = Enumerable.Repeat(2, maxColumns).ToArray();
            for (int r = 0; r < cellTextsArray.Length; r++)
            {
                for (int c = 0; c < maxColumns; c++)
                {
                    var lines = wrappedCells[r][c];
                    foreach (var line in lines)
                    {
                        targetWidths[c] = Math.Max(targetWidths[c], line.Length + 2);
                    }
                }
            }
        }

        state.Output.Append(MakeLine("┌", "┬", "┐", "┬", targetWidths));

        if (headerCells.Count > 0)
        {
            RenderRow(state, style, cellTextsArray, alignmentsArray, targetWidths, 0, null, wrappedCells);
        }

        state.Output.Append(MakeLine("├", "┼", "┤", "┼", targetWidths));

        for (int r = 0; r < dataRows.Count; r++)
        {
            var row = dataRows[r];
            var rowCells = row.Children ?? new List<MarkdownNode>();
            var rowAlignments = Enumerable.Range(0, maxColumns)
                .Select(c => c < rowCells.Count ? (rowCells[c].Info ?? "") : "")
                .ToArray();
            RenderRow(state, style, cellTextsArray, rowAlignments, targetWidths, r + 1, rowCells, wrappedCells);
        }

        state.Output.Append(MakeLine("└", "┴", "┘", "┴", targetWidths));
        state.Output.AppendLine();
    }

    private static void RenderRow(RendererState state, string style, string[][] cellTexts, string[] alignments, int[] targetWidths, int rowIdx, List<MarkdownNode>? rowCells, string[][][]? wrappedCells)
    {
        int maxColumns = targetWidths.Length;
        int maxLines = 1;
        string[][] rowLines = new string[maxColumns][];

        for (int c = 0; c < maxColumns; c++)
        {
            if (wrappedCells != null)
            {
                rowLines[c] = wrappedCells[rowIdx][c];
            }
            else
            {
                string text = rowIdx < cellTexts.Length && c < cellTexts[rowIdx].Length ? cellTexts[rowIdx][c] : "";
                rowLines[c] = [text];
            }
            maxLines = Math.Max(maxLines, rowLines[c].Length);
        }

        string align = "";
        if (rowCells != null && rowCells.Count > 0)
        {
            align = rowCells[0].Info ?? "";
        }
        else if (rowIdx == 0 && alignments.Length > 0)
        {
            align = alignments[0];
        }

        for (int lineIdx = 0; lineIdx < maxLines; lineIdx++)
        {
            state.Output.Append($"{style}│{Ansi.Reset}");
            for (int c = 0; c < maxColumns; c++)
            {
                string lineText = lineIdx < rowLines[c].Length ? rowLines[c][lineIdx] : "";
                string colAlign = c < alignments.Length ? alignments[c] : "";
                state.Output.Append($" {PadText(lineText, targetWidths[c] - 2, colAlign)}{style}│{Ansi.Reset}");
            }
            state.Output.Append('\n');
        }
    }

    private static int[] FitColumns(int[] columnWidths, int lineWidth, int numColumns, string[][] cellTexts)
    {
        int available = lineWidth - 1 - numColumns;
        int[] targetWidths = (int[])columnWidths.Clone();

        int[] minWidths = columnWidths.Select((_, colIdx) =>
        {
            int maxWordLen = 1;
            foreach (var row in cellTexts)
            {
                string text = colIdx < row.Length ? row[colIdx] : "";
                foreach (var word in text.Split(' '))
                {
                    maxWordLen = Math.Max(maxWordLen, word.Length);
                }
            }
            return maxWordLen + 2;
        }).ToArray();

        while (targetWidths.Sum() > available)
        {
            int maxIdx = 0;
            for (int i = 1; i < targetWidths.Length; i++)
            {
                if (targetWidths[i] > targetWidths[maxIdx]) maxIdx = i;
            }
            if (targetWidths[maxIdx] <= minWidths[maxIdx]) break;
            targetWidths[maxIdx]--;
        }

        return targetWidths;
    }

    private static string[][][] WrapAllCells(string[][] cellTexts, int[] targetWidths)
    {
        var result = new string[cellTexts.Length][][];
        for (int r = 0; r < cellTexts.Length; r++)
        {
            result[r] = new string[targetWidths.Length][];
            for (int c = 0; c < targetWidths.Length; c++)
            {
                string text = (r < cellTexts.Length && c < cellTexts[r].Length) ? cellTexts[r][c] : "";
                result[r][c] = WrapText(text, targetWidths[c] - 2);
            }
        }
        return result;
    }

    private static string[] WrapText(string text, int maxWidth)
    {
        if (text.Length <= maxWidth) return [text];
        var words = text.Split(' ');
        var lines = new List<string>();
        string currentLine = "";
        foreach (var word in words)
        {
            if (currentLine.Length == 0)
            {
                currentLine = word;
            }
            else if (currentLine.Length + 1 + word.Length <= maxWidth)
            {
                currentLine += " " + word;
            }
            else
            {
                lines.Add(currentLine);
                currentLine = word;
            }
        }
        if (currentLine.Length > 0) lines.Add(currentLine);
        return lines.ToArray();
    }

    private static string PadText(string text, int width, string align)
    {
        if (align == "right")
        {
            return new string(' ', width - text.Length) + text + " ";
        }
        if (align == "center")
        {
            var leftPad = (width - text.Length) / 2;
            var rightPad = width - text.Length - leftPad + 1;
            return new string(' ', leftPad) + text + new string(' ', rightPad);
        }
        return text + new string(' ', width - text.Length) + " ";
    }

    private static string GetTextFromNode(MarkdownNode node)
    {
        if (node.Type == "text")
        {
            return node.Content;
        }
        if (node.Children != null)
        {
            return string.Join("", node.Children.Select(GetTextFromNode));
        }
        return "";
    }
}

File: zig/src/render-console/consoleTableRenderer.zig
const std = @import("std");

const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const RendererState = @import("../types/RendererState.zig").RendererState;
const Renderer = @import("../types/Renderer.zig").Renderer;
const ansiDim = @import("console.zig").ansiDim;
const ansiReset = @import("console.zig").ansiReset;

pub const consoleTableRenderer = Renderer{
    .name = "table",
    .render = render,
};

pub fn render(node: *const MarkdownNode, state: *RendererState, decode: ?bool) void {
    _ = decode;

    if (state.output.items.len > 0 and state.output.items[state.output.items.len - 1] != '\n') {
        state.output.append(state.allocator, '\n') catch unreachable;
    }

    const children = node.children orelse return;

    if (children.len == 0) return;

    const headerRow = children[0];
    const dataRows = children[1..];

    const headerCells = headerRow.children orelse &.{};

    var maxColumns: usize = headerCells.len;
    for (dataRows) |row| {
        if (row.children) |rowCells| {
            if (rowCells.len > maxColumns) {
                maxColumns = rowCells.len;
            }
        }
    }

    var columnWidths = state.allocator.alloc(usize, maxColumns) catch unreachable;
    defer state.allocator.free(columnWidths);
    @memset(columnWidths, 0);

    var columnAlignments = state.allocator.alloc([]const u8, maxColumns) catch unreachable;
    defer state.allocator.free(columnAlignments);
    @memset(columnAlignments, "");

    var cellTexts = std.ArrayList([]const []const u8).initCapacity(state.allocator, dataRows.len + 1) catch unreachable;
    defer {
        for (cellTexts.items) |row| {
            for (row) |text| {
                state.allocator.free(text);
            }
            state.allocator.free(row);
        }
        cellTexts.deinit(state.allocator);
    }

    var headerTexts = state.allocator.alloc([]const u8, maxColumns) catch unreachable;
    @memset(headerTexts, "");
    cellTexts.append(state.allocator, headerTexts) catch unreachable;

    {
        var i: usize = 0;
        while (i < headerCells.len) : (i += 1) {
            const text = getTextFromNode(headerCells[i], state.allocator);
            headerTexts[i] = text;
            columnWidths[i] = @max(columnWidths[i], text.len + 2);
            columnAlignments[i] = headerCells[i].info orelse "";
        }
    }

    {
        var r: usize = 0;
        while (r < dataRows.len) : (r += 1) {
            const row = dataRows[r];
            const rowCells = row.children orelse &.{};
            const rowTexts = state.allocator.alloc([]const u8, maxColumns) catch unreachable;
            @memset(rowTexts, "");
            cellTexts.append(state.allocator, rowTexts) catch unreachable;

            var c: usize = 0;
            while (c < rowCells.len) : (c += 1) {
                const text = getTextFromNode(rowCells[c], state.allocator);
                rowTexts[c] = text;
                columnWidths[c] = @max(columnWidths[c], text.len + 2);
            }
        }
    }

    var targetWidths = columnWidths;
    var fitWidthsOwned: ?[]usize = null;
    var twOwned: ?[]usize = null;
    defer {
        if (fitWidthsOwned) |fw| state.allocator.free(fw);
        if (twOwned) |tw| state.allocator.free(tw);
    }

    if (state.line_width) |lineWidth| {
        const totalWidth: usize = 1 + sum(columnWidths) + maxColumns;
        if (totalWidth > lineWidth) {
            const fitWidths = fitColumns(state.allocator, columnWidths, lineWidth, maxColumns, cellTexts.items);
            fitWidthsOwned = fitWidths;

            var tw = state.allocator.alloc(usize, maxColumns) catch unreachable;
            twOwned = tw;
            @memset(tw, 2);

            for (0..cellTexts.items.len) |r| {
                for (0..maxColumns) |c| {
                    const text = if (c < cellTexts.items[r].len) cellTexts.items[r][c] else "";
                    const lines = wrapText(state.allocator, text, fitWidths[c] - 2);
                    defer freeWrappedLines(state.allocator, lines);
                    for (lines) |line| {
                        tw[c] = @max(tw[c], line.len + 2);
                    }
                }
            }
            targetWidths = tw;
        }
    }

    const makeLine = struct {
        fn func(left: []const u8, mid: []const u8, right: []const u8, s: *RendererState, widths: []const usize) !void {
            s.output.appendSlice(s.allocator, ansiDim) catch unreachable;
            s.output.appendSlice(s.allocator, left) catch unreachable;
            for (0..widths.len) |i| {
                var dashes = std.ArrayList(u8).initCapacity(s.allocator, widths[i]) catch unreachable;
                defer dashes.deinit(s.allocator);
                for (0..widths[i]) |_| {
                    dashes.appendSlice(s.allocator, "─") catch unreachable;
                }
                const dashStr = dashes.items;
                s.output.appendSlice(s.allocator, dashStr) catch unreachable;
                if (i < widths.len - 1) {
                    s.output.appendSlice(s.allocator, mid) catch unreachable;
                }
            }
            s.output.appendSlice(s.allocator, right) catch unreachable;
            s.output.appendSlice(s.allocator, ansiReset) catch unreachable;
            s.output.append(s.allocator, '\n') catch unreachable;
        }
    }.func;

    try makeLine("┌", "┬", "┐", state, targetWidths);

    if (headerCells.len > 0) {
        renderRow(state, cellTexts.items, 0, targetWidths, maxColumns, columnAlignments);
    }

    try makeLine("├", "┼", "┤", state, targetWidths);

    {
        var r: usize = 0;
        while (r < dataRows.len) : (r += 1) {
            renderRow(state, cellTexts.items, r + 1, targetWidths, maxColumns, columnAlignments);
        }
    }

    try makeLine("└", "┴", "┘", state, targetWidths);
}

fn renderRow(
    state: *RendererState,
    cellTexts: []const []const []const u8,
    rowIdx: usize,
    targetWidths: []const usize,
    maxColumns: usize,
    columnAlignments: []const []const u8,
) void {
    const allocator = state.allocator;

    var cellLines = allocator.alloc([]const []const u8, maxColumns) catch unreachable;
    defer allocator.free(cellLines);

    var maxLines: usize = 1;

    for (0..maxColumns) |c| {
        const text = if (rowIdx < cellTexts.len and c < cellTexts[rowIdx].len) cellTexts[rowIdx][c] else "";
        const lines = wrapText(allocator, text, targetWidths[c] - 2);
        cellLines[c] = lines;
        maxLines = @max(maxLines, lines.len);
    }
    defer {
        for (cellLines) |lines| {
            freeWrappedLines(allocator, lines);
        }
    }

    var line: usize = 0;
    while (line < maxLines) : (line += 1) {
        state.output.appendSlice(allocator, ansiDim) catch unreachable;
        state.output.appendSlice(allocator, "│") catch unreachable;
        state.output.appendSlice(allocator, ansiReset) catch unreachable;

        for (0..maxColumns) |c| {
            const text = if (line < cellLines[c].len) cellLines[c][line] else "";
            const alignment = columnAlignments[c];
            renderPaddedCell(state, text, targetWidths[c], alignment) catch unreachable;
        }
        state.output.append(allocator, '\n') catch unreachable;
    }
}

fn fitColumns(
    allocator: std.mem.Allocator,
    columnWidths: []const usize,
    lineWidth: usize,
    numColumns: usize,
    cellTexts: []const []const []const u8,
) []usize {
    const available = lineWidth - 1 - numColumns;
    var targetWidths = allocator.dupe(usize, columnWidths) catch unreachable;

    const minWidths = allocator.alloc(usize, columnWidths.len) catch unreachable;
    defer allocator.free(minWidths);

    for (0..columnWidths.len) |colIdx| {
        var maxWordLen: usize = 1;
        for (cellTexts) |row| {
            const text = if (colIdx < row.len) row[colIdx] else "";
            var start: usize = 0;
            for (text, 0..) |ch, i| {
                if (ch == ' ') {
                    const wordLen = i - start;
                    maxWordLen = @max(maxWordLen, wordLen);
                    start = i + 1;
                }
            }
            const wordLen = text.len - start;
            maxWordLen = @max(maxWordLen, wordLen);
        }
        minWidths[colIdx] = maxWordLen + 2;
    }

    while (sum(targetWidths) > available) {
        var maxIdx: usize = 0;
        for (1..targetWidths.len) |i| {
            if (targetWidths[i] > targetWidths[maxIdx]) maxIdx = i;
        }
        if (targetWidths[maxIdx] <= minWidths[maxIdx]) break;
        targetWidths[maxIdx] -= 1;
    }

    return targetWidths;
}

fn wrapText(allocator: std.mem.Allocator, text: []const u8, maxWidth: usize) []const []const u8 {
    if (text.len <= maxWidth) {
        var result = allocator.alloc([]const u8, 1) catch unreachable;
        result[0] = text;
        return result;
    }
    var words = std.ArrayList([]const u8).initCapacity(allocator, 8) catch unreachable;
    defer words.deinit(allocator);

    var start: usize = 0;
    for (text, 0..) |ch, i| {
        if (ch == ' ') {
            if (start < i) words.append(allocator, text[start..i]) catch unreachable;
            start = i + 1;
        }
    }
    if (start < text.len) words.append(allocator, text[start..]) catch unreachable;

    var lines = allocator.alloc([]const u8, words.items.len) catch unreachable;
    var lineCount: usize = 0;

    var currentLine = std.ArrayList(u8).initCapacity(allocator, maxWidth) catch unreachable;

    for (words.items) |word| {
        if (currentLine.items.len == 0) {
            currentLine.appendSlice(allocator, word) catch unreachable;
        } else if (currentLine.items.len + 1 + word.len <= maxWidth) {
            currentLine.append(allocator, ' ') catch unreachable;
            currentLine.appendSlice(allocator, word) catch unreachable;
        } else {
            lines[lineCount] = currentLine.toOwnedSlice(allocator) catch unreachable;
            lineCount += 1;
            currentLine = std.ArrayList(u8).initCapacity(allocator, maxWidth) catch unreachable;
            currentLine.appendSlice(allocator, word) catch unreachable;
        }
    }
    if (currentLine.items.len > 0) {
        lines[lineCount] = currentLine.toOwnedSlice(allocator) catch unreachable;
        lineCount += 1;
    } else {
        currentLine.deinit(allocator);
    }

    return allocator.realloc(lines, lineCount) catch unreachable;
}

fn freeWrappedLines(allocator: std.mem.Allocator, lines: []const []const u8) void {
    if (lines.len > 1) {
        for (lines) |line| {
            allocator.free(line);
        }
    }
    allocator.free(lines);
}

fn sum(widths: []const usize) usize {
    var total: usize = 0;
    for (widths) |w| total += w;
    return total;
}

fn getTextFromNode(node: *const MarkdownNode, allocator: std.mem.Allocator) []const u8 {
    if (std.mem.eql(u8, node.type, "text")) {
        return allocator.dupe(u8, node.content) catch unreachable;
    }
    if (node.children) |children| {
        var buffer = std.ArrayList(u8).initCapacity(allocator, 64) catch unreachable;
        for (children) |child| {
            if (std.mem.eql(u8, child.type, "text") or std.mem.eql(u8, child.type, "table_cell_content")) {
                buffer.appendSlice(allocator, child.content) catch unreachable;
            } else if (child.children) |grandchildren| {
                for (grandchildren) |grandchild| {
                    if (std.mem.eql(u8, grandchild.type, "text") or std.mem.eql(u8, grandchild.type, "table_cell_content")) {
                        buffer.appendSlice(allocator, grandchild.content) catch unreachable;
                    }
                }
            }
        }
        return buffer.toOwnedSlice(allocator) catch unreachable;
    }
    if (node.content.len > 0) {
        return allocator.dupe(u8, node.content) catch unreachable;
    }
    return "";
}

fn renderPaddedCell(state: *RendererState, text: []const u8, width: usize, alignment: []const u8) !void {
    state.output.append(state.allocator, ' ') catch unreachable;

    const innerWidth = width - 2;

    if (std.mem.eql(u8, alignment, "right")) {
        const padding = innerWidth - text.len;
        for (0..padding) |_| {
            state.output.append(state.allocator, ' ') catch unreachable;
        }
        state.output.appendSlice(state.allocator, text) catch unreachable;
        state.output.append(state.allocator, ' ') catch unreachable;
    } else if (std.mem.eql(u8, alignment, "center")) {
        const padding = innerWidth - text.len;
        const leftPad = padding / 2;
        const rightPad = padding - leftPad + 1;
        for (0..leftPad) |_| {
            state.output.append(state.allocator, ' ') catch unreachable;
        }
        state.output.appendSlice(state.allocator, text) catch unreachable;
        for (0..rightPad) |_| {
            state.output.append(state.allocator, ' ') catch unreachable;
        }
    } else {
        state.output.appendSlice(state.allocator, text) catch unreachable;
        const padding = innerWidth - text.len;
        for (0..padding) |_| {
            state.output.append(state.allocator, ' ') catch unreachable;
        }
        state.output.append(state.allocator, ' ') catch unreachable;
    }

    state.output.appendSlice(state.allocator, ansiDim) catch unreachable;
    state.output.appendSlice(state.allocator, "│") catch unreachable;
    state.output.appendSlice(state.allocator, ansiReset) catch unreachable;
}
